### PR TITLE
feat: Download original asset (DEV-3832)

### PR DESCRIPTION
**Pull request by siers**
implementation of `projects / shortcodePathVar / "assets" / path[AssetId]("assetId") / "original"` with the description: 
> Offers the original file for upload, provided the API permisisons allow.
> Authorization: JWT bearer token.

The handler calls DSP-API, disabled in tests. Route tested, the fetcher only through the docker setup.

### CODE REVIEW

File: src/main/scala/swiss/dasch/FetchAssetPermissions.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  ): Task[Int] =
+    (for {
+      uri <-
+        ZIO.succeed(
+          uri"http://${apiConfig.host}:${apiConfig.port}/admin/files/${assetInfo.assetRef.belongsToProject}/${assetInfo.derivative.filename}",
+        )
+      response    <- sttp.send(basicRequest.get(uri).header("Authorization", s"Bearer ${jwt}"))
+      successBody <- ZIO.fromEither(response.body).mapError(httpError(uri.toString, response.code.code, _))
+      permissionCode <-
+        ZIO.fromEither(successBody.fromJson[PermissionResponse].bimap(e => throw new Exception(e), _.permissionCode))
```
Comment (BalduinLandolt) [inline src/main/scala/swiss/dasch/FetchAssetPermissions.scala:47]: I'm not so familiar with the sttp client, and what the bimap function does here. But the `throw` looks suspicious here - if this actually throws the exception, then that will result in a `ZIO.die` not in a `ZIO.fail`, like the return type `Task[Int]` would suggest.  
Have you tested this error case and verifies that it behaves as expected?

File: src/test/scala/swiss/dasch/api/ProjectsEndpointSpec.scala [hunk 2]
```diff
@@ -167,6 +171,41 @@ object ProjectsEndpointSpec extends ZIOSpecDefault {
...
 
+  private val assetOriginalSuite =
+    suite("/projects/<shortcode>/asset/<assetId>/original")(
+      test("given the info file does not exist, it should return Not Found") {
+        val req = Request
+          .get(URL(Path.root / "projects" / "0666" / "assets" / "7l5QJAtPnv5-lLmBPfO7U40" / "original"))
+          .addHeader("Authorization", "Bearer fakeToken")
+        executeRequest(req).map(response => assertTrue(response.status == Status.NotFound))
+      },
+      test("return the original contents") {
```
Comment (BalduinLandolt) [inline src/test/scala/swiss/dasch/api/ProjectsEndpointSpec.scala:182]: would be nice to have a 3rd test case, for when the API denies access

**Review by BalduinLandolt — APPROVED** [review]
Over all looks good. I remarked one potential issue, in case the API denies access to the asset. Please let me know if it is a false alarm, or address, if it is indeed an issue.

File: src/main/scala/swiss/dasch/FetchAssetPermissions.scala (hunk 1 above)
Comment (siers) [reply src/main/scala/swiss/dasch/FetchAssetPermissions.scala:47]: Yeah, throw is not necessary here.

File: src/test/scala/swiss/dasch/api/ProjectsEndpointSpec.scala (hunk 2 above)
Comment (siers) [reply src/test/scala/swiss/dasch/api/ProjectsEndpointSpec.scala:187]: I'm having a problem with overriding the layer. I'll try to figure it out, but you can take a look.

Comment (seakayone) [reply src/test/scala/swiss/dasch/api/ProjectsEndpointSpec.scala:195]: You could change the `Mock` implementation to contain a `Ref` to the answer the specific tests desire.

Something like this:
```
class MockString(answer: Ref[String]){

  def getString(): UIO[String] = answer.get

  def setAnswerGetString(str: String) = answer.set(str)

}
```

And in the test you call `mock.setAnswerGetString("foo")` just before calling the method.

**Review by seakayone — CHANGES_REQUESTED** [review]

File: src/main/scala/swiss/dasch/api/ProjectsEndpoints.scala [hunk 3]
```diff
@@ -182,6 +186,17 @@ final case class ProjectsEndpoints(base: BaseEndpoints) {
     .tag("assets")
     .description("Authorization: read:project:1234 scope required.")
 
+  val getProjectsAssetsOriginal = base.secureEndpoint.get
+    .in(projects / shortcodePathVar / "assets" / path[AssetId]("assetId") / "original")
```
[the remark this answers is not in the document]
Comment (siers) [reply src/main/scala/swiss/dasch/api/ProjectsEndpoints.scala:190]: copypasted with minor differences, please take a look, reopen if needed

**Comment by siers** [comment]
Regarding stateful mocks, I was aware I could write it that way and intentionally didn't.

File: src/main/scala/swiss/dasch/FetchAssetPermissions.scala [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  apiConfig: Configuration.ApiConfig,
+) extends FetchAssetPermissions {
+  def getPermissionCode(
+    jwt: String,
+    assetInfo: AssetInfo,
+  ): Task[Int] =
+    (for {
+      uri <-
+        ZIO.succeed(
+          uri"http://${apiConfig.host}:${apiConfig.port}/admin/files/${assetInfo.assetRef.belongsToProject}/${assetInfo.derivative.filename}",
```
[the remark this answers is not in the document]
Comment (siers) [reply src/main/scala/swiss/dasch/FetchAssetPermissions.scala:42]: Isn't SSL terminated between containers?

File: src/main/resources/application.conf [hunk 5]
```diff
@@ -35,6 +35,13 @@ ingest {
   bulk-max-parallel = ${?INGEST_BULK_MAX_PARALLEL}
 }
 
+api {
+  host = localhost
+  host = ${?KNORA_WEBAPI_KNORA_API_EXTERNAL_HOST}
+  port = 3333
+  port = ${?KNORA_WEBAPI_KNORA_API_EXTERNAL_PORT}
```
[the remark this answers is not in the document]
Comment (siers) [reply src/main/resources/application.conf]: Implemented.

**Review by seakayone — APPROVED** [review]

File: src/main/scala/swiss/dasch/FetchAssetPermissions.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (seakayone) [reply src/main/scala/swiss/dasch/FetchAssetPermissions.scala]: Yes it is. I was assuming we will configure the external url.

**Pull request merged**